src/librustdoc/html/render/mod.rs

```diff
@@ -2718,6 +2730,10 @@ fn item_trait(w: &mut Buffer, cx: &Context, it: &clean::Item, t: &clean::Trait,
             }
         }
 
+        let mut implementors = implementors.clone();
```

Comment:
It's kinda useless to sort `implementors` since it's not used for rendering as is. Look at `synthetic` and `concrete` (and update the `compare_impl` function too like I said above).

Reply:
But I wanted to sort foreign types. Why is it bad to sort `implementors`? Aren’t synthetic and concrete created it out of it?

Reply:
The problem is that synthetic and concrete are later re-sorted by `compare_impl` (lines 2744 and 2745 in this diff, or https://github.com/rust-lang/rust/blob/6df26f897cffb2d86880544bb451c6b5f8509b2d/src/librustdoc/html/render/print_item.rs#L656-L667).

You need a sort function that compares first by polarity, and then by compare_impl. You should sort implementors by that, and then remove the subsequent .sort_by calls on `synthetic` and `concrete` so you don't undo the sorting.